core/grant/service.go

```diff
@@ -254,6 +256,34 @@ func (s *Service) Revoke(ctx context.Context, id, actor, reason string, opts ...
 	return grant, nil
 }
 
+func (s *Service) Restore(ctx context.Context, id, actor, reason string) (*domain.Grant, error) {
+	grant, err := s.GetByID(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("getting grant details: %w", err)
+	}
+
+	if err := grant.Restore(actor, reason); err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrInvalidRequest, err.Error())
+	}
+
+	if err := s.providerService.GrantAccess(ctx, *grant); err != nil {
+		return nil, fmt.Errorf("granting access in provider: %w", err)
+	}
+
+	if err := s.repo.Update(ctx, grant); err != nil {
+		return nil, fmt.Errorf("updating grant record in db: %w", err)
+	}
```

Comment:
What happen if Grant has been granted then repo Update fails? Will it be an issue?

Reply:
yeah need to handle the rollback, will update this